### PR TITLE
Move Throughput into `sc-sysinfo`

### PR DESCRIPTION
**Pull request by Szegoo**
Prepares for #12017

Polkadot address: 126X27SbhrV19mBFawys3ovkyBS87SGfYwtwa8J2FjHrtbmA

### CODE REVIEW

**Comment by ggwpez** [comment]
I think it is also fine to split the issue into two MRs. The changes here already make sense to have on their own.

**Comment by Szegoo** [comment]
> I think it is also fine to split the issue into two MRs, depending on your preferences. The changes here already make sense to have on their own.

Sure, that makes sense.

**Comment by ggwpez** [comment]
I can review earliest on Friday. But maybe @koute wants to, it touches your code :)   

You are only reafactoring code as preparation for #12017, right?  
Then please update the issue description to not *Close* the issue.  
We also need to ensure that this does not break telemetry.

**Comment by Szegoo** [comment]
> You are only reafactoring code as preparation for #12017, right? Then please update the issue description to not _Close_ the issue. We also need to ensure that this does not break telemetry.

Yes, this is only preparation for #12017, I changed the description now. The CI doesn't seem happy, I will do some minor fixes on this.

**Comment by bkchr** [comment]
I also don't really get why we need `Throughput` there. @ggwpez can you give more context?

File: client/sysinfo/src/sysinfo.rs [hunk 1]
```diff
@@ -32,6 +33,77 @@ use std::{
...
+/// Throughput as measured in bytes per second.
+#[derive(Deserialize, Serialize, Debug, Clone, Copy, PartialEq)]
+pub enum Throughput {
+	/// KiB/s
+	KiBs(f64),
+	/// MiB/s
+	MiBs(f64),
+	/// GiB/s
+	GiBs(f64),
+}
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:45]: Hm... now that I look at it I find having this as an enum a little strange; can we refactor this to be a normal newtype?

Simply put, something like this:

```
pub struct Throughput(f64);
```

This would keep the results at full bytes/s precision.

Also, please note that either way this requires either a) a custom implementation of `Serialize`, or annotating the `HwBench` structure with a `#[serde(serialize_with = "...")]` to preserve the format which the telemetry expects (which is always in MB/s). Since this is going to hold the full precision of values yet the serialization is only in MB/s it feels more natural to me to go with the option B.

@ggwpez Thoughts?

Comment (bkchr) [reply client/sysinfo/src/sysinfo.rs:45]: That is a very good idea! And then the `Display` implementation could show the best fitting formatting depending on the size of the value.

Comment (koute) [reply client/sysinfo/src/sysinfo.rs:45]: Yes.

Comment (ggwpez) [reply client/sysinfo/src/sysinfo.rs:45]: I chose that enum originally to make use of serde to have nice JSON formats eg:  
```json
"MiBs": 450.0
```
But we can probably re-build that through the serde deserialization, or?

Comment (koute) [reply client/sysinfo/src/sysinfo.rs:45]: > But we can probably re-build that through the serde deserialization, or?

Yes. For where we need a pretty output for human consumable JSON we can just easily attach a custom serializer, so that shouldn't be a problem.

**Comment by ggwpez** [comment]
> I also don't really get why we need `Throughput` there. @ggwpez can you give more context?

The measurement of the performance at startup runs in `sc-sysinfo`. So if we now want to compare the results of that to anything, we need the `Throughput` types to have nicer compare logic and formatting in the final output.  
I just originally did not put it in `sy-sysinfo` to keep the `sc-` crate small, since it was not needed.

File: client/sysinfo/src/sysinfo.rs (hunk 1 above)
Comment (ggwpez) [reply client/sysinfo/src/sysinfo.rs:45]: We could also consider moving the `Metric` type, but that would only be for uniform naming.

File: client/sysinfo/src/sysinfo.rs [hunk 2]
```diff
@@ -32,6 +33,64 @@ use std::{
...
+#[derive(Deserialize, Serialize, Debug, Clone, Copy, PartialEq)]
+pub struct Throughput(f64);
+
+const KIBIBYTE: f64 = 1024.0;
+
+impl Throughput {
+	/// `f64` kibibyte/s to byte/s.
+	pub fn from_kibs(mibs: f64) -> Throughput {
+		Throughput(mibs * KIBIBYTE)
+	}
```
Comment (Szegoo) [inline client/sysinfo/src/sysinfo.rs:46]: Updated the `Throughput`, does this look reasonable?

Comment (koute) [reply client/sysinfo/src/sysinfo.rs:46]: Yeah, it's better.

Few more things from the top of my head that can/should be improved:
- Telemetry's still broken.
- You should modify the `benchmark` function so that it returns a value with full precision. (And it can probably just return `Throughput`, I guess?)
- I'd rename the `to_kibs`/etc. functions to `as_kibs`/etc.. (That's how e.g. `std::time::Duration` names its functions; let's be consistent)
- The `to_kibs`/etc. should not return a `Throughput` anymore; `Throughput` internally should *always* be the same unit (that is, in bytes)

File: client/sysinfo/src/sysinfo.rs [hunk 3]
```diff
@@ -58,9 +121,9 @@ pub(crate) fn benchmark<E>(
 		}
 	}
 
-	let score = ((size * count) as f64 / elapsed.as_secs_f64()) / (1024.0 * 1024.0);
+	let score = Throughput((size * count) as f64 / elapsed.as_secs_f64());
 	log::trace!(
-		"Calculated {} of {:.2}MB/s in {} iterations in {}ms",
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:63]: The `B/s` should be unnecessary since the `Display` impl should automatically pick the unit now, no?

File: client/sysinfo/src/sysinfo.rs [hunk 4]
```diff
@@ -121,7 +184,7 @@ pub const DEFAULT_CPU_EXECUTION_LIMIT: ExecutionLimit =
 	ExecutionLimit::Both { max_iterations: 4 * 1024, max_duration: Duration::from_millis(100) };
 
 // This benchmarks the CPU speed as measured by calculating BLAKE2b-256 hashes, in MB/s.
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:186]:
```suggestion
// This benchmarks the CPU speed as measured by calculating BLAKE2b-256 hashes, in bytes per second.
```

Same for all the other mentions where it previously said that it was in MB/s.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 5]
```diff
@@ -177,15 +106,15 @@ mod tests {
...
+		assert_eq_error_rate_float!(14667.776 * 1024.0, gib.as_kibs(), EPS);
 		assert_eq!("14.32 GiB/s", gib.to_string());
-		assert_eq!("14.32 GiB/s", gib.normalize().to_string());
+		assert_eq!("14.32 GiB/s", gib.normalize().0.to_string());
 
-		let mib = Throughput::MiBs(1029.0);
+		let mib = Throughput::from_mibs(1029.0);
 		assert_eq!("1.00 GiB/s", mib.to_string());
 	}
 }
```
Comment (koute) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:120]: This test should probably be also moved, since it tests `Throughput` and we moved `Throughput`.

File: client/sysinfo/src/sysinfo.rs [hunk 6]
```diff
@@ -32,14 +33,76 @@ use std::{
...
+		self.0 / (KIBIBYTE * KIBIBYTE)
+	}
+
+	/// [`Self`] as number of gibibyte/s.
+	pub fn as_gibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// Normalizes [`Self`] to use the larges unit possible.
+	fn normalize(&self) -> (f64, &'static str) {
```
Comment (Szegoo) [inline client/sysinfo/src/sysinfo.rs:78]: I don't think we need the `normalize` function to be public anymore since it is only used for formatting.

Comment (koute) [reply client/sysinfo/src/sysinfo.rs:78]: Yeah, if it's not used outside of this crate then sounds good to me.

File: client/sysinfo/src/sysinfo.rs [hunk 7]
```diff
@@ -32,14 +33,98 @@ use std::{
...
+}
+
+impl<'de> Deserialize<'de> for Throughput {
+    fn deserialize<D>(deserializer: D) -> Result<Throughput, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+		Ok(Throughput(1 as f64))
+    }
+}
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:147]: 1. The `Serialize` should be implemented as a standalone function and attached to the fields of `HwBench` with `#[serde(serialize_with = "path")]` (since it requires special handling of serialization where it's always in megabytes)
2. I don't think we actually need deserialization. (:

File: client/sysinfo/src/lib.rs [hunk 8]
```diff
@@ -44,13 +44,13 @@ pub const TARGET_ENV: &str = include_str!(concat!(env!("OUT_DIR"), "/target_env.
...
 	/// Memory bandwidth in MB/s, calculated by measuring the throughput of `memcpy`.
-	pub memory_memcpy_score: u64,
+	pub memory_memcpy_score: Throughput,
 	/// Sequential disk write speed in MB/s.
-	pub disk_sequential_write_score: Option<u64>,
+	pub disk_sequential_write_score: Option<Throughput>,
 	/// Random disk write speed in MB/s.
-	pub disk_random_write_score: Option<u64>,
+	pub disk_random_write_score: Option<Throughput>,
 }
```
Comment (koute) [inline client/sysinfo/src/lib.rs:54]: Could you please also add a test which serializes this structure into JSON and compares it to a preset string? (Since the format needs to explicitly match what telemetry expects we should have a test that enforces it.)

File: client/sysinfo/src/sysinfo.rs (hunk 7 above)
Comment (Szegoo) [reply client/sysinfo/src/sysinfo.rs:147]: This is old code, I updated it now. Are you sure we don't need deserialization? Without custom deserialization the code didn't compile, now that I added deserialization it should.

Comment (Szegoo) [reply client/sysinfo/src/sysinfo.rs]: Should the `HwBench` always be in megabytes?

Comment (koute) [reply client/sysinfo/src/sysinfo.rs]: I think so. At least not generic deserialization. Some of @ggwpez 's code might need it, but it's probably more appropriate to handle it the same way as with the serialization here (that is, with an attribute). Can you point to the exact piece of code which doesn't compile when you remove deserialization?

Comment (koute) [reply client/sysinfo/src/sysinfo.rs]: > Should the `HwBench` always be in megabytes?

Yes.

File: client/sysinfo/src/lib.rs (hunk 8 above)
Comment (Szegoo) [reply client/sysinfo/src/lib.rs:54]: https://github.com/paritytech/substrate/blob/952030cfa6f11be6aef938e5359064c4cf6b30a9/utils/frame/benchmarking-cli/src/machine/hardware.rs#L34
This didn't work without the custom deserialization.

File: client/sysinfo/src/sysinfo.rs (hunk 7 above)
Comment (Szegoo) [reply client/sysinfo/src/sysinfo.rs]: https://github.com/paritytech/substrate/blob/952030cfa6f11be6aef938e5359064c4cf6b30a9/utils/frame/benchmarking-cli/src/machine/hardware.rs#L34

This didn't work without the custom deserialization.

Comment (koute) [reply client/sysinfo/src/sysinfo.rs]: Ah yes, this is what @ggwpez was talking about. That should just use a custom deserializer attached with an attribute, just like the `HwBench`.

File: client/sysinfo/src/sysinfo.rs [hunk 9]
```diff
@@ -32,14 +37,144 @@ use std::{
...
+		} else if key == "MiBs" {
+			return Ok(Throughput(value * KIBIBYTE * KIBIBYTE))
+		} else {
+			return Ok(Throughput(value * KIBIBYTE * KIBIBYTE * KIBIBYTE))
+		}
+	}
+}
+
+impl<'de> Deserialize<'de> for Throughput {
+	fn deserialize<D>(deserializer: D) -> Result<Throughput, D::Error>
```
Comment (Szegoo) [inline client/sysinfo/src/sysinfo.rs:162]: I believe that it is the best to have the deserialization be implemented for `Throughput` instead of a stand-alone function, since the deserialization logic is always the same.

**Comment by Szegoo** [comment]
Should I move `Metric` also? Or should that be done in a separate PR?

File: client/sysinfo/src/sysinfo.rs (hunk 9 above)
Comment (koute) [reply client/sysinfo/src/sysinfo.rs:162]: My reasoning here follows something like this:

1. The natural way to serialize/deserialize this is to just do it in full precision. (Which is what a `#[derive]`'d impl would do.)
2. Telemetry doesn't use this and needs a different serialization format. (Not full precision; in MB/s)
3. The code in benchmarking-cli needs a different serialization format. (Wrapped with the unit for readability)
4. Since nothing uses the "natural" serialization format and every piece of code expects a different serialization format in isn't appropriate to define it on the type itself (it's unnecessary, and one might use it by mistake)
5. If the serialization isn't defined on the type itself then neither the deserialization should.

So basically just move the `Deserialize` impl into a standalone function, and move it into the benchmarking-cli where it's actually used.

File: client/sysinfo/src/sysinfo.rs [hunk 10]
```diff
@@ -450,32 +586,63 @@ mod tests {
...
+		let hwbench = HwBench {
+			cpu_hashrate_score: Throughput::from_gibs(1.32),
+			memory_memcpy_score: Throughput::from_kibs(1342.432),
+			disk_sequential_write_score: Some(Throughput::from_kibs(432.12)),
+			disk_random_write_score: None,
+		};
+
+		let serialized = serde_json::to_string(&hwbench).unwrap();
+		// All the throughput should be converted to MiBs.
+		assert_eq!(serialized, "{\"cpu_hashrate_score\":{\"MiBs\":1351.68},\"memory_memcpy_score\":{\"MiBs\":1.31096875},\"disk_sequential_write_score\":{\"MiBs\":0.4219921875},\"disk_random_write_score\":{}}");
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:646]: 1. These should be serialized as bare values, without the `{}`.
2. These should be serialized as integers since the telemetry expects integers. (e.g. `1351.68` should be `1351`)

**Comment by koute** [comment]
> Should I move `Metric` also? Or should that be done in a separate PR?

Let's do that in a separate PR.

File: client/sysinfo/Cargo.toml [hunk 11]
```diff
@@ -26,3 +26,4 @@ sc-telemetry = { version = "4.0.0-dev", path = "../telemetry" }
 sp-core = { version = "6.0.0", path = "../../primitives/core" }
 sp-io = { version = "6.0.0", path = "../../primitives/io" }
 sp-std = { version = "4.0.0", path = "../../primitives/std" }
+sp-runtime = { version = "6.0.0", path = "../../primitives/runtime" }
```
Comment (koute) [inline client/sysinfo/Cargo.toml:29]: This should be a dev dependency since it's only required in the tests.

File: client/sysinfo/src/sysinfo.rs [hunk 12]
```diff
@@ -450,32 +581,63 @@ mod tests {
...
+		let hwbench = HwBench {
+			cpu_hashrate_score: Throughput::from_gibs(1.32),
+			memory_memcpy_score: Throughput::from_kibs(9342.432),
+			disk_sequential_write_score: Some(Throughput::from_kibs(4332.12)),
+			disk_random_write_score: None,
+		};
+
+		let serialized = serde_json::to_string(&hwbench).unwrap();
+		// All the throughput should be converted to MiBs.
+		assert_eq!(serialized, "{\"cpu_hashrate_score\":1351,\"memory_memcpy_score\":9,\"disk_sequential_write_score\":4,\"disk_random_write_score\":0}");
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:641]: When a value is `None` it should not be serialized.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 13]
```diff
@@ -45,6 +45,7 @@ pub struct Requirement {
 	/// The metric to measure.
 	pub metric: Metric,
 	/// The minimal throughput that needs to be archived for this requirement.
+	#[serde(serialize_with = "serialize_throughput", deserialize_with = "deserialize_throughput")]
```
Comment (koute) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:48]: These serialization/deserialization functions should just be moved to this file (since this is the only thing that needs those)

File: client/sysinfo/src/sysinfo.rs (hunk 12 above)
Comment (Szegoo) [reply client/sysinfo/src/sysinfo.rs:594]: Do you mean by this that `disk_random_write_score\":0` should be `disk_random_write_score\":null` instead? Or do you mean that we have nothing serialized here because I am not sure how we should go about that?

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 14]
```diff
@@ -18,8 +18,55 @@
...
+	Deserialize, Deserializer, Serialize, Serializer,
+};
+use sp_std::{fmt, fmt::Formatter};
+
+pub fn serialize_throughput<S>(t: &Throughput, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	let mut map = serializer.serialize_map(Some(1))?;
+	let (value, unit) = t.normalize();
```
Comment (koute) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:34]: Hmm.... I'm a little spooked by using this here. It's not obvious in the `sysinfo` crate that the exact string returned from `normalize` is used for serialization/deserialization here (especially since over there it's only used for formatting), so I can imagine someone could e.g. change how the units are called or the formatting slightly over there and it'd break serialization here.

Could we maybe change `normalize` to return an *enum* instead? e.g. get it to return something like this instead of a string:

```
enum Unit {
    GiBs,
    MiBs,
    KiBs
}
```

File: client/sysinfo/src/sysinfo.rs [hunk 15]
```diff
@@ -32,14 +33,113 @@ use std::{
...
+
+pub fn serialize_throughput_option_as_mibs<S>(
+	maybe_t: &Option<Throughput>,
+	serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	if let Some(t) = maybe_t {
+		return serializer.serialize_u64(t.as_mibs() as u64)
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:130]: Nitpick: in case of json it doesn't change anything, but this should be using `serialize_some`, as shown in the example snippet to which I linked to in my previous comment.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 16]
```diff
@@ -18,8 +18,58 @@
...
+	map.serialize_entry(&unit.to_string(), &value)?;
```
Comment (koute) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:35]: As I've said in the previous comment, the problem here was that it was not obvious in the `sysinfo` crate that the exact string returned from `normalize` is used here for serialization/deserialization.

You changed it to use an enum (which is good!), but you still kept the "enum -> serialization key" conversion exactly where it was previously (in the `sysinfo` crate). The whole point of making this into an enum was so that we can move that conversion here so that 1) how it's printed, and 2) how it's serialized here are kept entirely independent. That way if someone were to modify either one of them in the future they won't inadvertently break the other one. No spooky action at a distance.

So this should not use `to_string` on the enum; just do a `match` on the enum instead.

File: client/sysinfo/src/sysinfo.rs [hunk 17]
```diff
@@ -32,14 +33,113 @@ use std::{
...
+}
+
+impl fmt::Display for Unit {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let unit = match self {
+			Unit::GiBs => "GiBs",
+			Unit::MiBs => "MiBs",
+			Unit::KiBs => "KiBs",
+		};
+		write!(f, "{}", unit)
```
Comment (koute) [inline client/sysinfo/src/sysinfo.rs:49]: You can just use `f.write_str` here.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs (hunk 16 above)
Comment (koute) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs]: It's fine either way. The whole point of `Display` is, well, displaying something to the user, and that's what we're doing in `sc-sysinfo`.

(Note: if something implements `Display` then you don't need to `.to_string()`; you can just print it directly.)

Comment (Szegoo) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs]: Other than the nit you mentioned do the changes look good now?

**Review by koute — APPROVED** [review]

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 18]
```diff
@@ -18,8 +18,64 @@
...
+};
+use sp_std::{fmt, fmt::Formatter};
+
+pub fn serialize_throughput<S>(t: &Throughput, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	let mut map = serializer.serialize_map(Some(1))?;
+	let (value, unit) = t.normalize();
+	let unit_as_str = match unit {
```
Comment (ggwpez) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:35]:
```suggestion
	let unit_as_str = format!("{}", unit);
```
An `Unit` should implement `Display` IIUC.

File: client/sysinfo/src/sysinfo.rs [hunk 19]
```diff
@@ -32,14 +33,117 @@ use std::{
 	time::{Duration, Instant},
 };
 
+/// Used to represent the unit used in the benchmarks.
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:36]:
```suggestion
/// The unit in which the [`Throughput`] (bytes per second) is denoted. 
```
`Unit` is a bit weak as name, but its not a blocker from my side.

File: client/sysinfo/src/sysinfo.rs [hunk 20]
```diff
@@ -32,14 +33,117 @@ use std::{
...
+
+/// Throughput as measured in bytes per second.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub struct Throughput(f64);
+
+const KIBIBYTE: f64 = 1024.0;
+
+impl Throughput {
+	/// `f64` kibibyte/s to byte/s.
+	pub fn from_kibs(kibs: f64) -> Throughput {
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:61]:
```suggestion
	/// Construct [`Self`] from kibibyte/s.
	pub fn from_kibs(kibs: f64) -> Throughput {
```
... analogous for the functions below.

File: client/sysinfo/src/sysinfo.rs [hunk 21]
```diff
@@ -32,14 +33,117 @@ use std::{
...
+		Throughput(mibs * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// `f64` gibibyte/s to byte/s.
+	pub fn from_gibs(gibs: f64) -> Throughput {
+		Throughput(gibs * KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// [`Self`] as f64.
+	pub fn as_f64(&self) -> f64 {
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:76]:
```suggestion
	/// [`Self`] as number of byte/s.
	pub fn as_byte(&self) -> f64 {
```

File: client/sysinfo/src/sysinfo.rs [hunk 22]
```diff
@@ -32,14 +33,117 @@ use std::{
...
+			Unit::KiBs => "KiBs",
+		})
+	}
+}
+
+/// Throughput as measured in bytes per second.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub struct Throughput(f64);
+
+const KIBIBYTE: f64 = 1024.0;
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:57]:
```suggestion
const KIBIBYTE: f64 = (1<<10) as f64;
const MEBIBYTE: f64 = (1<<20) as f64;
const GIBIBYTE: f64 = (1<<30) as f64;
```
Then you have to do less float operations later.

File: client/sysinfo/src/sysinfo.rs [hunk 23]
```diff
@@ -32,14 +33,117 @@ use std::{
...
+}
+
+impl fmt::Display for Throughput {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let (value, unit) = self.normalize();
+		write!(f, "{:.2?} {}", value, unit)
+	}
+}
+
+pub fn serialize_throughput_as_mibs<S>(
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:116]: All public functions and types need some docs.

File: client/sysinfo/src/sysinfo.rs [hunk 24]
```diff
@@ -58,9 +162,9 @@ pub(crate) fn benchmark<E>(
 		}
 	}
 
-	let score = ((size * count) as f64 / elapsed.as_secs_f64()) / (1024.0 * 1024.0);
+	let score = Throughput((size * count) as f64 / elapsed.as_secs_f64());
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:165]: Why are you just constructing it when we have these constructors: `from_mebibyte` etc.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 25]
```diff
@@ -18,8 +18,64 @@
...
-use std::fmt;
+use sc_sysinfo::{Throughput, Unit};
+use serde::{
+	de::{Error, MapAccess, Visitor},
+	ser::SerializeMap,
+	Deserialize, Deserializer, Serialize, Serializer,
+};
+use sp_std::{fmt, fmt::Formatter};
+
+pub fn serialize_throughput<S>(t: &Throughput, serializer: S) -> Result<S::Ok, S::Error>
```
Comment (ggwpez) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:29]: This is only needed for serde? Then please add a comment.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 26]
```diff
@@ -18,8 +18,64 @@
...
+	{
+		let (key, value): (&str, f64) = if let Some((key, value)) = access.next_entry()? {
+			(key, value)
+		} else {
+			return Err(M::Error::custom("Expected an entry."))
+		};
+
+		match key {
+			"KiBs" => Ok(Throughput::from_kibs(value)),
+			"MiBs" => Ok(Throughput::from_mibs(value)),
```
Comment (ggwpez) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:66]: Could implement this on the `Unit`, or?

File: utils/frame/benchmarking-cli/src/machine/hardware.rs (hunk 18 above)
Comment (ggwpez) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs:35]: Talked to @koute its fine.

File: client/sysinfo/src/sysinfo.rs (hunk 21 above)
Comment (Szegoo) [reply client/sysinfo/src/sysinfo.rs:76]: Considering the other function names maybe it would be better to call this `as_bytes`, or not?

Comment (ggwpez) [reply client/sysinfo/src/sysinfo.rs]: :+1:

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 27]
```diff
@@ -45,6 +101,7 @@ pub struct Requirement {
 	/// The metric to measure.
 	pub metric: Metric,
 	/// The minimal throughput that needs to be archived for this requirement.
+	#[serde(serialize_with = "serialize_throughput", deserialize_with = "deserialize_throughput")]
```
[the remark this answers is not in the document]
Comment (bkchr) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs:104]: Ahh, now I see. Do we really need to serialize with the unit attached? I don't think we should do this. You can just serialize as mib and fine.

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 28]
```diff
@@ -18,8 +18,64 @@
...
+	let mut map = serializer.serialize_map(Some(1))?;
+	let (value, unit) = t.normalize();
+	let unit_as_str = match unit {
+		Unit::GiBs => "GiBs",
+		Unit::MiBs => "MiBs",
+		Unit::KiBs => "KiBs",
+	};
+
+	map.serialize_entry(&unit_as_str, &value)?;
+	map.end()
```
[the remark this answers is not in the document]
Comment (koute) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs]: @bkchr I explicitly requested that this will be done the way it is now. (:

See these comments why:

https://github.com/paritytech/substrate/pull/12368#discussion_r986833804
https://github.com/paritytech/substrate/pull/12368#discussion_r988985571

Maybe a comment in the code would be good why this `match` is currently duplicated? Or just don't serialize/deserialize the units at all and keep it all in MB/s (but that's up to @ggwpez).

File: utils/frame/benchmarking-cli/src/machine/hardware.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (Szegoo) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs:104]: I don't think we should do this, the `reference_hardware.json` has values specified in other units also, of course, we can probably change that, but isn't it nicer to have the biggest possible unit used?

File: utils/frame/benchmarking-cli/src/machine/hardware.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs:42]: Same argument as above - just the maintainability of the json file. Since having units is self explanatory, otherwise it needs docs to explain what it means.

[the remark this answers is not in the document]
Comment (bkchr) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs:42]: Hmm okay, I'm not really convinced of having the mibs etc be part of the json. However, I also don't really care. If @ggwpez is happy, he can approve and I will not complain.

File: client/sysinfo/src/sysinfo.rs [hunk 29]
```diff
@@ -58,9 +166,9 @@ pub(crate) fn benchmark<E>(
 		}
 	}
 
-	let score = ((size * count) as f64 / elapsed.as_secs_f64()) / (1024.0 * 1024.0);
+	let score = Throughput::from_mibs((size * count) as f64 / elapsed.as_secs_f64());
```
Comment (ggwpez) [inline client/sysinfo/src/sysinfo.rs:169]:
```suggestion
	let score = Throughput::from_kibs((size * count) as f64 / (elapsed.as_secs_f64() / 1024.0));
```
since it is returned in bytes/s.

**Review by ggwpez — APPROVED** [review]
I dont have a strong opinion on the serialization stuff.

**Comment by Szegoo** [comment]
@ggwpez So what is the final decision then?

**Comment by ggwpez** [comment]
> @ggwpez So what is the final decision then?

Okay then lets remove the unit. Less code - less bugs.

**Comment by Szegoo** [comment]
> Okay then lets remove the unit from the serialization. Less code - less bugs.

@ggwpez What should we do with `reference_hardware.json` then?

**Comment by ggwpez** [comment]
> @ggwpez What should we do with `reference_hardware.json` then?

Change it so that it works with the new format.

> Edit: wouldn't auto-generating it result in even more code?

We need auto-generation at one point anyway, as manual editing is error-prone and not straight forward.

**Comment by Szegoo** [comment]
@ggwpez So I assume we should use MiBs in the `reference_hardware.json` then?

**Comment by bkchr** [comment]
@ggwpez can we merge this?

File: utils/frame/benchmarking-cli/src/machine/hardware.rs [hunk 30]
```diff
@@ -31,6 +63,7 @@ lazy_static! {
 	/// <https://wiki.polkadot.network/docs/maintain-guides-how-to-validate-polkadot>
 	pub static ref SUBSTRATE_REFERENCE_HARDWARE: Requirements = {
 		let raw = include_bytes!("reference_hardware.json").as_slice();
+		println!("{:?}", raw);
```
Comment (ggwpez) [inline utils/frame/benchmarking-cli/src/machine/hardware.rs:66]:
```suggestion
```

Comment (Szegoo) [reply utils/frame/benchmarking-cli/src/machine/hardware.rs]: This was by accident 😅

**Comment by ggwpez** [comment]
bot merge
